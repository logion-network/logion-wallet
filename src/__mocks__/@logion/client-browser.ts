export class BrowserFile {

    constructor(blob: Blob) {
        this.blob = blob;
    }

    private blob: Blob;

    getBlob() {
        return this.blob;
    }
}

export class BrowserAxiosFileUploader {

    async upload() {

    }
}
