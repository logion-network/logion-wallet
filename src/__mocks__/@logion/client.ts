import { CheckResultType } from "@logion/client/dist/CollectionItem.js";
import { DefaultSignAndSendStrategy } from "@logion/client/dist/Signer.js";

import {
    LogionClient,
    AccountTokens,
    toIsoString,
    fromIsoString,
    ClosedCollectionLoc,
    DraftRequest,
    EditableRequest,
    OpenLoc,
    LocRequestState,
    LocsState,
    HashOrContent,
} from '../LogionClientMock';

export {
    LogionClient,
    AccountTokens,
    toIsoString,
    fromIsoString,
    CheckResultType,
    ClosedCollectionLoc,
    DraftRequest,
    EditableRequest,
    OpenLoc,
    DefaultSignAndSendStrategy,
    LocRequestState,
    LocsState,
    HashOrContent,
}

export { MimeType } from "@logion/client/dist/Mime.js";
