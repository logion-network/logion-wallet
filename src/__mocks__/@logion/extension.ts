import {
    enableExtensions,
    ExtensionSigner,
    isExtensionAvailable,
    enableMetaMask,
    allMetamaskAccounts,
} from '../PolkadotExtensionDappMock';
export {
    enableExtensions,
    ExtensionSigner,
    isExtensionAvailable,
    enableMetaMask,
    allMetamaskAccounts,
};
