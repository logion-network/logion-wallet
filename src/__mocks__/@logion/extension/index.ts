import {
    enableExtensions,
    ExtensionSigner,
    isExtensionAvailable,
    enableMetaMask,
    allMetamaskAccounts,
} from '../../LogionExtensionMock';
export {
    enableExtensions,
    ExtensionSigner,
    isExtensionAvailable,
    enableMetaMask,
    allMetamaskAccounts,
};
