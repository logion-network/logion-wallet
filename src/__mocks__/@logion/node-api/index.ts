import { ApiPromise } from "src/__mocks__/PolkadotApiMock";
export * from "@logion/node-api/dist/UUID";

export function buildApi() {
    return new ApiPromise();
}

export * from "./dist/LogionLoc";
