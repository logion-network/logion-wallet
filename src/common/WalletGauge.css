.WalletGauge .actions {
    display: flex;
    justify-content: center;
}

.WalletGauge .actions button:nth-child(n+2) {
    margin-left: 25px;
}

.WalletGauge {
    display: flex;
    align-items: flex-end;
    justify-content: space-between;
}

.WalletGauge .actions {
    margin-bottom: 10px;
    flex-shrink: 0;
    height: 36px;
}

.TransferDialog .EstimatedFees {
    margin-top: 20px;
    margin-bottom: 20px;
}
