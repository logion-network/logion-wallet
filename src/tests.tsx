import { ReactElement } from 'react';
import { createRenderer } from 'react-test-renderer/shallow';
import reactRenderer, { ReactTestRenderer, ReactTestRendererJSON, act as reactAct } from 'react-test-renderer';
import { screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

export function render(element: ReactElement): null | ReactTestRendererJSON | ReactTestRendererJSON[] {
    const component = reactRenderer.create(element);
    return component.toJSON();
}

export function renderer(element: ReactElement): ReactTestRenderer {
    return reactRenderer.create(element);
}

export function shallowRender<E extends ReactElement>(element: ReactElement): E {
    const renderer = createRenderer();
    renderer.render(element);
    return renderer.getRenderOutput();
}

export function mockAccount(address: string, name: string) {
    return {
        address,
        meta: {
            name
        }
    };
}

export { reactAct as act };

export async function clickByName(name: string | RegExp | ((accessibleName: string, element: Element) => boolean)) {
    let button: HTMLElement;
    await waitFor(() => button = screen.getByRole("button", { name }));
    await userEvent.click(button!);
}

export async function typeByLabel(label: string, value: string) {
    let input: HTMLElement;
    await waitFor(() => input = screen.getByLabelText(label));
    await userEvent.type(input!, value);
}
