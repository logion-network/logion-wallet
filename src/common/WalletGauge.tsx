import { useState, useCallback, useMemo, useEffect } from 'react';
import { Form, Spinner, InputGroup, DropdownButton, Dropdown } from 'react-bootstrap';
import { BalanceState } from '@logion/client';
import { Numbers, CoinBalance, Lgnt } from '@logion/node-api';

import { CallCallback, useLogionChain } from '../logion-chain';

import Gauge from './Gauge';
import Button from './Button';
import Icon from './Icon';
import Dialog from './Dialog';
import FormGroup from './FormGroup';
import { ExpectNewTransactionStatus, useCommonContext } from './CommonContext';
import Alert from './Alert';

import './WalletGauge.css';
import BalanceDetails from './BalanceDetails';
import ExtrinsicSubmissionStateView from 'src/ExtrinsicSubmissionStateView';

export interface Props {
    balance: CoinBalance,
    type: 'arc' | 'linear',
    vaultAddress?: string,
    sendButton?: boolean
}

interface TransferDialogState {
    title: string;
    destination: boolean;
    show: boolean;
}

const HIDDEN_DIALOG: TransferDialogState = {
    title: "",
    destination: true,
    show: false,
};

export default function WalletGauge(props: Props) {
    const { accounts, signer, client, submitCall, extrinsicSubmissionState, clearSubmissionState } = useLogionChain();
    const { colorTheme, mutateBalanceState, expectNewTransaction, expectNewTransactionState, stopExpectNewTransaction } = useCommonContext();
    const [ destination, setDestination ] = useState("");
    const [ amount, setAmount ] = useState("");
    const [ unit, setUnit ] = useState(Numbers.NONE);
    const [ transferDialogState, setTransferDialogState ] = useState<TransferDialogState>(HIDDEN_DIALOG);
    const { vaultAddress, sendButton } = props;

    const transfer = useMemo(() => {
        return async (callback: CallCallback) => {
            await mutateBalanceState(async (state: BalanceState) => {
                return await state.transfer({
                    amount: Lgnt.fromPrefixedNumber(new Numbers.PrefixedNumber(amount, unit)),
                    destination,
                    callback,
                    signer: signer!
                });
            });
            expectNewTransaction();
        };
    }, [ amount, destination, unit, mutateBalanceState, signer, expectNewTransaction ]);

    const transferCallback = useCallback(async () => {
        try {
            await submitCall(transfer);
        } catch(e) {}
    }, [ transfer, submitCall ]);

    const clearFormCallback = useCallback(() => {
        setDestination("");
        setAmount("");
        setTransferDialogState(HIDDEN_DIALOG);
    }, [ setDestination, setAmount ]);

    const cancelCallback = useCallback(() => {
        clearFormCallback();
        stopExpectNewTransaction();
        clearSubmissionState();
    }, [ clearFormCallback, stopExpectNewTransaction, clearSubmissionState ]);

    useEffect(() => {
        if(expectNewTransactionState.status === ExpectNewTransactionStatus.DONE) {
            cancelCallback();
        }
    }, [ expectNewTransactionState, cancelCallback ]);

    if(!client) {
        return null;
    }

    return (
        <div className={ "WalletGauge " + props.type }>
            <Gauge
                readingIntegerPart={ props.balance.available.coefficient.toInteger() }
                readingDecimalPart={ props.balance.available.coefficient.toFixedPrecisionDecimals(2) }
                unit={ props.balance.available.prefix.symbol + Lgnt.CODE }
                level={ props.balance.level }
                type={ props.type }
            />
            <BalanceDetails
                balance={ props.balance }
                type={ props.type }
            />
            { (sendButton === undefined || sendButton) &&
            <div className="actions">
                    <Button slim onClick={ () => {
                        setTransferDialogState({
                            title: `Transfer ${ Lgnt.CODE }s`,
                            destination: true,
                            show: true,
                        });
                    } }>
                        <Icon icon={ { id: 'send' } } /> Send
                    </Button>
                { vaultAddress !== undefined &&
                    <Button slim onClick={ () => {
                        setDestination(vaultAddress);
                        setTransferDialogState({
                            title: `Transfer ${ Lgnt.CODE }s to your logion Vault`,
                            destination: false,
                            show: true,
                        });
                    } }>
                        <Icon icon={ { id: 'vault-in' } } /> Send to Vault
                    </Button>
                }
            </div>
            }
            <Dialog
                show={ transferDialogState.show }
                actions={ [
                    {
                        id: 'cancel',
                        buttonText: "Cancel",
                        buttonVariant: 'secondary-polkadot',
                        callback: cancelCallback,
                        disabled: extrinsicSubmissionState.inProgress,
                    },
                    {
                        id: 'transfer',
                        buttonText: "Transfer",
                        buttonVariant: 'polkadot',
                        callback: transferCallback,
                        disabled: !extrinsicSubmissionState.canSubmit() || !client.isValidAddress(destination) || isNaN(Number(amount)) || Number(amount) === 0 || destination === accounts!.current!.accountId.address
                    }
                ] }
                size="lg"
            >
                <h3>{ transferDialogState.title }</h3>
                {
                    expectNewTransactionState.status === ExpectNewTransactionStatus.IDLE &&
                    <>
                        { transferDialogState.destination &&
                            <FormGroup
                                id="destination"
                                label="Destination"
                                control={ <Form.Control
                                    isInvalid={ destination !== "" && (!client.isValidAddress(destination) || destination === accounts!.current!.accountId.address) }
                                    type="text"
                                    placeholder="The beneficiary's SS58 address"
                                    value={ destination }
                                    onChange={ value => setDestination(value.target.value) }
                                /> }
                                colors={ colorTheme.dialog }
                            />
                        }
                        <FormGroup
                            id="amount"
                            label="Amount"
                            noFeedback={ true }
                            control={
                                <InputGroup hasValidation>
                                    <Form.Control
                                        isInvalid={ amount !== "" && isNaN(Number(amount)) }
                                        type="text"
                                        placeholder="The amount to transfer"
                                        value={ amount }
                                        onChange={ value => setAmount(value.target.value) }
                                    />
                                    <DropdownButton
                                        title={ `${ unit.symbol }${ Lgnt.CODE }` }
                                        id="input-group-dropdown-1"
                                    >{
                                        [
                                            Numbers.NONE,
                                            Numbers.MILLI,
                                            Numbers.MICRO,
                                            Numbers.NANO,
                                            Numbers.PICO,
                                            Numbers.FEMTO,
                                            Numbers.ATTO
                                        ].map(unit => <Dropdown.Item key={ unit.symbol }
                                                                        onClick={ () => setUnit(unit) }>{ `${ unit.symbol }${ Lgnt.CODE }` }</Dropdown.Item>)
                                    }</DropdownButton>
                                    <Form.Control.Feedback type="invalid">
                                        Please enter a valid amount.
                                    </Form.Control.Feedback>
                                </InputGroup>
                            }
                            colors={ colorTheme.dialog }
                        />
                        <ExtrinsicSubmissionStateView
                            successMessage="Transfer successful."
                        />
                    </>
                }
                {
                    expectNewTransactionState.status === ExpectNewTransactionStatus.WAITING_NEW_TRANSACTION &&
                    <Alert variant="info">
                        <Spinner animation="border" />
                        <p>Transfer successful, waiting for the transaction to be finalized.</p>
                        <p>Note that this may take up to 30 seconds. If you want to proceed, you can safely
                            click on cancel but your transaction may not show up yet.</p>
                    </Alert>
                }
            </Dialog>
        </div>
    );
}
