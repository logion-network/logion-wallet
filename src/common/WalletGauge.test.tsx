jest.mock("../logion-chain");

import { shallowRender } from '../tests';
import { CoinBalance, Queries, Numbers } from '@logion/node-api';

import WalletGauge from './WalletGauge';

describe("WalletGauge", () => {

    it("renders", () => {
        const balance: CoinBalance = {
            coin: Queries.getCoin('lgnt'),
            available: new Numbers.PrefixedNumber("20.00", Numbers.MILLI),
            reserved: new Numbers.PrefixedNumber("0", Numbers.NONE),
            total: new Numbers.PrefixedNumber("20.00", Numbers.MILLI),
            level: 0.5,
        };
        const result = shallowRender(<WalletGauge
            balance={ balance }
            sendButton={ true }
            sendToVault={ false }
            withdrawFromVault={ false }
        />);
        expect(result).toMatchSnapshot();
    });    
});
