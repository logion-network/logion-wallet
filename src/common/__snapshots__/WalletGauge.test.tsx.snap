// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders arc 1`] = `
<div
  className="WalletGauge arc"
>
  <Gauge
    level={0.5}
    readingDecimalPart="00"
    readingIntegerPart="20"
    type="arc"
    unit="mLGNT"
  />
  <BalanceDetails
    balance={
      Object {
        "available": PrefixedNumber {
          "_prefix": Object {
            "symbol": "m",
            "tenExponent": -3,
          },
          "_scientificNumber": ScientificNumber {
            "_normalized": NormalizedNumber {
              "_decimalPart": "",
              "_integerPart": "20",
              "_negative": false,
              "_normalized": "20.",
            },
            "_tenExponent": -3,
          },
        },
        "coin": Object {
          "id": "lgnt",
          "symbol": "LGNT",
        },
        "level": 0.5,
        "reserved": PrefixedNumber {
          "_prefix": Object {
            "symbol": "",
            "tenExponent": 0,
          },
          "_scientificNumber": ScientificNumber {
            "_normalized": NormalizedNumber {
              "_decimalPart": "",
              "_integerPart": "",
              "_negative": false,
              "_normalized": ".",
            },
            "_tenExponent": 0,
          },
        },
        "total": PrefixedNumber {
          "_prefix": Object {
            "symbol": "m",
            "tenExponent": -3,
          },
          "_scientificNumber": ScientificNumber {
            "_normalized": NormalizedNumber {
              "_decimalPart": "",
              "_integerPart": "20",
              "_negative": false,
              "_normalized": "20.",
            },
            "_tenExponent": -3,
          },
        },
      }
    }
    type="arc"
  />
  <div
    className="actions"
  >
    <Button
      onClick={[Function]}
      slim={true}
    >
      <Icon
        icon={
          Object {
            "id": "send",
          }
        }
      />
       Send
    </Button>
  </div>
  <Dialog
    actions={
      Array [
        Object {
          "buttonText": "Cancel",
          "buttonVariant": "secondary-polkadot",
          "callback": [Function],
          "disabled": undefined,
          "id": "cancel",
        },
        Object {
          "buttonText": "Transfer",
          "buttonVariant": "polkadot",
          "callback": [Function],
          "disabled": true,
          "id": "transfer",
        },
      ]
    }
    show={false}
    size="lg"
  >
    <h3>
      
    </h3>
    <React.Fragment>
      <FormGroup
        colors={
          Object {
            "background": "#0c163d",
            "borderColor": "#e6007a",
            "foreground": "#ffffff",
          }
        }
        control={
          <FormControl
            isInvalid={false}
            onChange={[Function]}
            placeholder="The beneficiary's SS58 address"
            type="text"
            value=""
          />
        }
        id="destination"
        label="Destination"
      />
      <FormGroup
        colors={
          Object {
            "background": "#0c163d",
            "borderColor": "#e6007a",
            "foreground": "#ffffff",
          }
        }
        control={
          <InputGroup
            hasValidation={true}
          >
            <FormControl
              isInvalid={false}
              onChange={[Function]}
              placeholder="The amount to transfer"
              type="text"
              value=""
            />
            <DropdownButton
              id="input-group-dropdown-1"
              title="LGNT"
            >
              <DropdownItem
                onClick={[Function]}
              >
                LGNT
              </DropdownItem>
              <DropdownItem
                onClick={[Function]}
              >
                mLGNT
              </DropdownItem>
              <DropdownItem
                onClick={[Function]}
              >
                µLGNT
              </DropdownItem>
              <DropdownItem
                onClick={[Function]}
              >
                nLGNT
              </DropdownItem>
              <DropdownItem
                onClick={[Function]}
              >
                pLGNT
              </DropdownItem>
              <DropdownItem
                onClick={[Function]}
              >
                fLGNT
              </DropdownItem>
              <DropdownItem
                onClick={[Function]}
              >
                aLGNT
              </DropdownItem>
            </DropdownButton>
            <Feedback
              type="invalid"
            >
              Please enter a valid amount.
            </Feedback>
          </InputGroup>
        }
        id="amount"
        label="Amount"
        noFeedback={true}
      />
      <ExtrinsicSubmissionStateView
        successMessage="Transfer successful."
      />
    </React.Fragment>
  </Dialog>
</div>
`;

exports[`renders linear 1`] = `
<div
  className="WalletGauge linear"
>
  <Gauge
    level={0.5}
    readingDecimalPart="00"
    readingIntegerPart="20"
    type="linear"
    unit="mLGNT"
  />
  <BalanceDetails
    balance={
      Object {
        "available": PrefixedNumber {
          "_prefix": Object {
            "symbol": "m",
            "tenExponent": -3,
          },
          "_scientificNumber": ScientificNumber {
            "_normalized": NormalizedNumber {
              "_decimalPart": "",
              "_integerPart": "20",
              "_negative": false,
              "_normalized": "20.",
            },
            "_tenExponent": -3,
          },
        },
        "coin": Object {
          "id": "lgnt",
          "symbol": "LGNT",
        },
        "level": 0.5,
        "reserved": PrefixedNumber {
          "_prefix": Object {
            "symbol": "",
            "tenExponent": 0,
          },
          "_scientificNumber": ScientificNumber {
            "_normalized": NormalizedNumber {
              "_decimalPart": "",
              "_integerPart": "",
              "_negative": false,
              "_normalized": ".",
            },
            "_tenExponent": 0,
          },
        },
        "total": PrefixedNumber {
          "_prefix": Object {
            "symbol": "m",
            "tenExponent": -3,
          },
          "_scientificNumber": ScientificNumber {
            "_normalized": NormalizedNumber {
              "_decimalPart": "",
              "_integerPart": "20",
              "_negative": false,
              "_normalized": "20.",
            },
            "_tenExponent": -3,
          },
        },
      }
    }
    type="linear"
  />
  <div
    className="actions"
  >
    <Button
      onClick={[Function]}
      slim={true}
    >
      <Icon
        icon={
          Object {
            "id": "send",
          }
        }
      />
       Send
    </Button>
  </div>
  <Dialog
    actions={
      Array [
        Object {
          "buttonText": "Cancel",
          "buttonVariant": "secondary-polkadot",
          "callback": [Function],
          "disabled": undefined,
          "id": "cancel",
        },
        Object {
          "buttonText": "Transfer",
          "buttonVariant": "polkadot",
          "callback": [Function],
          "disabled": true,
          "id": "transfer",
        },
      ]
    }
    show={false}
    size="lg"
  >
    <h3>
      
    </h3>
    <React.Fragment>
      <FormGroup
        colors={
          Object {
            "background": "#0c163d",
            "borderColor": "#e6007a",
            "foreground": "#ffffff",
          }
        }
        control={
          <FormControl
            isInvalid={false}
            onChange={[Function]}
            placeholder="The beneficiary's SS58 address"
            type="text"
            value=""
          />
        }
        id="destination"
        label="Destination"
      />
      <FormGroup
        colors={
          Object {
            "background": "#0c163d",
            "borderColor": "#e6007a",
            "foreground": "#ffffff",
          }
        }
        control={
          <InputGroup
            hasValidation={true}
          >
            <FormControl
              isInvalid={false}
              onChange={[Function]}
              placeholder="The amount to transfer"
              type="text"
              value=""
            />
            <DropdownButton
              id="input-group-dropdown-1"
              title="LGNT"
            >
              <DropdownItem
                onClick={[Function]}
              >
                LGNT
              </DropdownItem>
              <DropdownItem
                onClick={[Function]}
              >
                mLGNT
              </DropdownItem>
              <DropdownItem
                onClick={[Function]}
              >
                µLGNT
              </DropdownItem>
              <DropdownItem
                onClick={[Function]}
              >
                nLGNT
              </DropdownItem>
              <DropdownItem
                onClick={[Function]}
              >
                pLGNT
              </DropdownItem>
              <DropdownItem
                onClick={[Function]}
              >
                fLGNT
              </DropdownItem>
              <DropdownItem
                onClick={[Function]}
              >
                aLGNT
              </DropdownItem>
            </DropdownButton>
            <Feedback
              type="invalid"
            >
              Please enter a valid amount.
            </Feedback>
          </InputGroup>
        }
        id="amount"
        label="Amount"
        noFeedback={true}
      />
      <ExtrinsicSubmissionStateView
        successMessage="Transfer successful."
      />
    </React.Fragment>
  </Dialog>
</div>
`;
